everviz/pages/objectives.py

```diff
@@ -1,41 +1,91 @@
 import os
 from collections import namedtuple
+from functools import partial
 import pandas as pd
+import numpy as np
 
-DataSources = namedtuple("DataSource", "objective_values")
+from everviz.log import get_logger
 
+DataSources = namedtuple(
+    "DataSource", ["objective_values", "objective_statistics", "total_objective_values"]
+)
 
-def _objective_values(api):
+logger = get_logger()
+
+
+def _objective_values_from_api(api):
     return pd.DataFrame(api.objective_values)
 
 
-def _calc_p10_p90(df):
-    percentiles = [("P10", 0.1), ("P90", 0.9)]
-    for batch in df["batch"].unique():
-        for label, quantile in percentiles:
-            df.loc[df["batch"] == batch, label] = df[df["batch"] == batch][
-                "value"
-            ].quantile(quantile)
-    return df
+def _total_objective_values_from_api(api):
+    return pd.DataFrame(api.single_objective_values)
+
+
+def _objective_values(data):
+    # Sort by the index columns.
+    data = data.drop(columns=["realization"])
+    sorted_values = (
+        data.set_index(["function", "batch", "simulation"]).sort_index().reset_index()
+    )
+    return sorted_values
+
+
+def _total_objective_values(data):
+    # Sort by the index columns.
+    sorted_values = data.set_index("batch").sort_index().reset_index()
+    return sorted_values
 
 
-def _calc_mean(df):
-    for batch in df["batch"].unique():
-        df.loc[df["batch"] == batch, "Mean"] = df[df["batch"] == batch]["value"].mean()
-    return df
+def _objective_statistics(data):
+    # Aggregate the values over the simulations, using pivot_table, keeping the
+    # batch and function as the multi-index, calculating statistics of the
+    # values over the simulations.
+    statistics = pd.pivot_table(
```

Comment:
Isn´t this almost identical to the statistics you made for summary keys? Could we use the same function for both?

Reply:
Yes, it is the same code logic, but they are not quite the same, since the shapes of the input data frames is different. A shared function would be either more complicated and harder to  understand, or only separate out the very basic function calls. I think it is better to leave it as it is.